### PR TITLE
Prevent stack overflow with `do` tactic

### LINKED ISSUE #15962: Stack overflow from "do 600000 idtac."
Ex: `Goal True. do 600000 idtac. (* "Stack overflow." *)`

Non-recursive loops shouldn't need lots of stack frames.  It works for 500000.  I don't know if it's system-dependent.

Version: 8.15.2+ (master)

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #15962

### CODE REVIEW

**Comment by Alizter** [comment]
Test?

**Comment by SkySkimmer** [comment]
Stack overflow is annoying to reproduce, we wouldn't know if the test is incorrectly passing.

**Comment by SkySkimmer** [comment]
Maybe a unit test would work.

**Comment by ppedrot** [comment]
Yet another example where partial application kills, btw.

File: tactics/tacticals.ml [hunk 1]
```diff
@@ -224,7 +224,10 @@ let tclDO n t =
     if k < 0 then user_err
       (str"Wrong argument : Do needs a positive integer.");
     if Int.equal k 0 then tclIDTAC
-    else if Int.equal k 1 then t else (tclTHEN t (dorec (k-1)))
+    else if Int.equal k 1 then t
```
Comment (ppedrot) [inline tactics/tacticals.ml:227]: Don't fix this one, it's dead code in master. It's scheduled for removal as soon as 8.16 is branched.

**Comment by Alizter** [comment]
Can you not do some trickery with ulimit to get a consistent stackoverflow?

File: tactics/tacticals.ml (hunk 1 above)
Comment (SkySkimmer) [reply tactics/tacticals.ml:227]: That's no reason to revert the fix imo

Comment (ppedrot) [reply tactics/tacticals.ml:227]: The rebase will be non-trivial, but if you insist...

**Comment by ppedrot** [comment]
I don't think we need a test, they're indeed annoying for this category of issues. The CI failures are spurious, so @coqbot merge

**Comment by ppedrot** [comment]
@coqbot merge now

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by jfehrle** [comment]
With this fix, I've discovered the following quickly exhausts available memory (e.g. 20 GB in a minute or so).  Could you take a look at it?  You can probably make progress far quicker than I can.  Thanks.

```
Set Ltac Debug.
Goal True. do 60000000 idtac.
```

**Pull request merged**